homeassistant/components/flashforge/config_flow.py

```diff
@@ -0,0 +1,150 @@
+
+        ip = None
+        port = 8899
+        local_ip = await async_get_source_ip(self.hass)
+        discovered_printers = await Discovery.getPrinters(
+            self.hass.loop,
+            limit=1,
+            host_ip=local_ip,
+        )
+        # Get the first discovered printer ip
+        for _, ip_addr in discovered_printers:
+            ip = ip_addr
+            break
+
+        if ip is None:
+            return self.async_abort(reason="no_devices_found")
+
+        try:
+            await self._get_printer_info(
+                self.hass, {CONF_IP_ADDRESS: ip, CONF_PORT: port}
+            )
+        except CannotConnect:
+            return self.async_abort(reason="no_devices_found")
+
+        self._set_confirm_only()
+        return self.async_show_form(
+            step_id="auto_confirm",
+            description_placeholders={
+                "machine_name": self.printer.machine_name,
+                "ip_addr": ip,
+            },
+        )
+
+    async def async_step_auto_confirm(self, _: dict[str, Any] = None) -> FlowResult:
+        """User confirmed to add device to Home Assistant."""
+        return self._async_create_entry()
+
+    @callback
+    def _async_show_form(
+        self,
+        errors: dict[str, str] = None,
+    ) -> FlowResult:
+        """Create and show the form for user."""
+
+        data_schema = vol.Schema(
+            {
+                vol.Optional(
+                    CONF_IP_ADDRESS,
+                    description={"suggested_value": self.ip_addr},
+                ): str,
+                vol.Optional(CONF_PORT, default=self.port): cv.port,
+            }
+        )
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=data_schema,
+            errors=errors or {},
+        )
+
+    async def _get_printer_info(self, hass: HomeAssistant, user_input: dict) -> None:
+        """Try to get info from given ip."""
+
+        self.ip_addr = user_input[CONF_IP_ADDRESS]
+        self.port = user_input[CONF_PORT]
+        self.printer = Printer(self.ip_addr, self.port)
+
+        try:
+            await self.printer.connect()
+        except TimeoutError as err:
+            raise CannotConnect(err) from err
+        except ConnectionError as err:
+            raise CannotConnect(err) from err
+
+        if self.printer.serial is not None:
+            await self.async_set_unique_id(self.printer.serial)
+
+        self._abort_if_unique_id_configured(
+            updates={CONF_IP_ADDRESS: self.ip_addr, CONF_PORT: self.port}
+        )
+
+    @callback
+    def _async_create_entry(self) -> FlowResult:
+        return self.async_create_entry(
+            title=self.printer.machine_name,
+            data={
+                CONF_IP_ADDRESS: self.ip_addr,
+                CONF_PORT: self.port,
+                CONF_SERIAL_NUMBER: self.printer.serial,
+            },
+        )
+
+
+class CannotConnect(exceptions.HomeAssistantError):
```

Comment:
There is no need to implement this one if you catch the exceptions like (TimeoutError and ConnectionError) above and just set  `errors["base"] = "cannot_connect"` in that situation.

Reply:
Would you find it better to skip the try/except block completely inside _get_printer_info and catch (TimeoutError, ConnectionError) when calling _get_printer_info ?